include/busclique/util.hpp

```diff
@@ -150,57 +151,44 @@ class ignore_badmask {};
     const size_t shore;
     const uint64_t seed;
     
-    topo_spec_base(size_t d0, size_t d1, size_t s, uint64_t e) :
-        dim{d0, d1}, shore(s), seed(e) {}
+    topo_spec_base(size_t dim_y, size_t dim_x, size_t s, uint64_t e) :
+        dim_y(dim_y), dim_x(dim_x), shore(s), seed(e) {}
 
-    topo_spec_base(size_t d0, size_t d1, size_t s, uint32_t e) :
-        topo_spec_base(d0, d1, s, fastrng::amplify_seed(e)) {}
+    topo_spec_base(size_t dim_y, size_t dim_x, size_t s, uint32_t e) :
+        topo_spec_base(dim_y, dim_x, s, fastrng::amplify_seed(e)) {}
 
     template<typename serialize_tag, typename ...Args>
     size_t serialize(serialize_tag, uint8_t *output, const Args &...args) const {
-        return _serialize(serialize_tag{}, output, dim, shore, seed, args...);
+        return _serialize(serialize_tag{}, output, dim_y, dim_x, shore, seed, args...);
     }
 
-    inline size_t cell_addr(size_t u, size_t y, size_t x) const {
-        return x + dim[1]*(y + dim[0]*u);
+    inline size_t cell_addr(size_y y, size_x x, bool u) const {
+        return coordinate_converter::cell_addr(y, x, u, dim_y, dim_x);
     }
 
-    inline size_t num_cells() const {
-        return dim[0]*dim[1]*2;
+    inline size_t cell_addr(bool u, size_w w, size_z z) const {
+        if (u) return cell_addr(horz(w), horz(z), u);
+        else   return cell_addr(vert(z), vert(w), u);
     }
 
-    inline void linemajor(size_t q, size_t &u, size_t &w, size_t &z, size_t &k) const {
-        k = q % shore;
-        linemajor(q, u, w, z);
-    }
-    inline void linemajor(size_t q, size_t &u, size_t &w, size_t &z) const {
-        size_t p[2];
-                   q /= shore;
-        u = q % 2; q /= 2;
-        p[0] = q % dim[1];
-        p[1] = q / dim[1];
-        w = p[u];
-        z = p[1-u];
+    inline size_t num_cells() const {
+        return coordinate_converter::product(dim_y, dim_x, 2);
     }
 
-    inline size_t chimera_linear(size_t y, size_t x, size_t u, size_t k) const {
-        minorminer_assert(u < 2);
-        minorminer_assert(k < shore);
-        minorminer_assert(y < dim[0]);
-        minorminer_assert(x < dim[1]);
-        return k + shore*(u + 2*(x + dim[1]*y));
+    template<typename shore_t>
+    inline size_t chimera_linear(size_y y, size_x x, bool u, shore_t k) const {
+        return coordinate_converter::chimera_linear(y, x, u, k, dim_y, dim_x, shore_t(shore));
     }
 
-    inline void chimera_coordinates(size_t q,
-                                    size_t &y, size_t &x, size_t &u, size_t &k) const {
-        k = q % shore;  q /= shore;
-        u = q % 2;      q /= 2;
-        x = q % dim[1]; y  = q/dim[1];
-        minorminer_assert(chimera_linear(y, x, u, k) == q);
+    template<typename shore_t>
+    inline void linear_chimera(size_t q, 
+                               size_y &y, size_x &x, bool &u, shore_t &k) const {
+        coordinate_converter::linear_chimera(q, y, x, u, k, dim_y, dim_x, shore_t(shore));
```

Comment:
For my own understanding: how come these functions aren't in chimera_spec_base instead?

Reply:
It's also used in `topo_cache`, `pegasus_spec_base` and `zephyr_spec_base` in handling of `badmask`